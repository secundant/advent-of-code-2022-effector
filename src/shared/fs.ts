import { readFile } from 'node:fs/promises';

export const readRelativeFile = (meta: ImportMeta, path: string) =>
  readFile(new URL(path, meta.url).pathname, 'utf-8');
