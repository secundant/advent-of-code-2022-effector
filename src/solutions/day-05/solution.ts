import { combine, createStore } from 'effector';

export interface MoveCommand {
  size: number;
  from: number;
  to: number;
}

const move = (stacks: string[][], { to, size, from }: MoveCommand) => {
  stacks[to].push(...stacks[from].splice(stacks[from].length - size).reverse());
};

export const $stacks = createStore<string[][]>([], { sid: '$stacks' });
export const $commands = createStore<MoveCommand[]>([], { sid: '$commands' });

export const $reorderedStacks = combine($stacks, $commands, (stacks, commands) =>
  commands.reduce(
    (acc, command) => {
      move(acc, command);
      return acc;
    },
    Array.from(stacks, stack => Array.from(stack))
  )
);

export const $reorderedTopCrates = $reorderedStacks.map(stacks =>
  stacks.map(stack => stack.at(-1))
);
export const $reorderedMask = $reorderedTopCrates.map(chars =>
  chars
    .filter(Boolean)
    .map(crateName => crateName!.slice(1, 2))
    .join('')
);
