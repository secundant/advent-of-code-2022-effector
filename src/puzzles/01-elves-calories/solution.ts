import { createEvent, createStore, sample } from 'effector';
import { max, sum } from '@/shared/math';

export const elvesWroteNewStatistics = createEvent<string>();

const $detailedCalories = createStore<number[][]>([]);

export const $maxCalories = $detailedCalories.map(stats => max(stats.map(sum)));

sample({
  clock: elvesWroteNewStatistics.map(text =>
    text
      .trim()
      .split('\n\n')
      .map(chunk => chunk.split('\n').map(line => Number.parseInt(line, 10)))
  ),
  target: $detailedCalories
});
