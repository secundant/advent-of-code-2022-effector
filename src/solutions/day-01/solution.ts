import { createEvent, createStore, sample } from 'effector';
import { sum } from '@/shared/math';
import { last, numericComparator, sort } from '@/shared/std';

const countOfElvesCarryingTheMostCalories = 3;

export const replaceCalories = createEvent<number[][]>();

const $rawCaloriesGroups = createStore<number[][]>([]);

const $calories = $rawCaloriesGroups.map(groups => groups.map(sum));
const $sortedCalories = $calories.map(calories => sort(calories, numericComparator));

export const $sumOfTopCalories = $sortedCalories.map(calories =>
  sum(calories.slice(-1 * countOfElvesCarryingTheMostCalories))
);
export const $maxCalories = $sortedCalories.map(last);

sample({
  clock: replaceCalories,
  target: $rawCaloriesGroups
});
