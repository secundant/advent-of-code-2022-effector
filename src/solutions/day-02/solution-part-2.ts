import { sum } from '@/shared/math';
import type { Encrypted, EncryptedRes, GameShape, Outcome } from '@/solutions/day-02/solution';
import {
  $encryptedRounds,
  outcomeScore,
  reqCode,
  shapes,
  shapeScore
} from '@/solutions/day-02/solution';

type RoundOutcome = [GameShape, Outcome];

const loseShapes = Object.fromEntries(
  Object.entries(shapes).map(([key, value]) => [value, key])
) as Record<GameShape, GameShape>;
const resAsOutcome: Record<EncryptedRes, Outcome> = { X: 'lose', Y: 'draw', Z: 'win' };

const decryptAsOutcome = ([req, res]: Encrypted): RoundOutcome => [reqCode[req], resAsOutcome[res]];
const restoreRoundScore = ([req, outcome]: RoundOutcome) => {
  const res = outcome === 'lose' ? shapes[req] : outcome === 'draw' ? req : loseShapes[req];

  return shapeScore[res] + outcomeScore[outcome];
};

export const $roundsOutcome = $encryptedRounds.map(rounds => rounds.map(decryptAsOutcome));
export const $restoredRoundsScore = $roundsOutcome.map(rounds => rounds.map(restoreRoundScore));
export const $restoredTotalScore = $restoredRoundsScore.map(sum);
