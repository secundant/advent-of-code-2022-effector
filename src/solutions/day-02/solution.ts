import { createStore } from 'effector';
import { sum } from '@/shared/math';

type Outcome = keyof typeof outcomeScore;
type GameShape = keyof typeof shapes;
type GameRound = [GameShape, GameShape];
type Encrypted = [keyof typeof reqCode, keyof typeof resCode];

const shapes = {
  rock: 'scissors',
  paper: 'rock',
  scissors: 'paper'
} as const;

const shapeScore = { rock: 1, paper: 2, scissors: 3 };
const outcomeScore = { lose: 0, draw: 3, win: 6 } satisfies Record<string, number>;

const reqCode = { A: 'rock', B: 'paper', C: 'scissors' } satisfies Record<string, GameShape>;
const resCode = { X: 'rock', Y: 'paper', Z: 'scissors' } satisfies Record<string, GameShape>;

const decrypt = ([req, res]: Encrypted): GameRound => [reqCode[req], resCode[res]];
const isWon = ([req, res]: GameRound) => shapes[res] === req;
const getRoundOutcome = ([req, res]: GameRound): Outcome =>
  isWon([req, res]) ? 'win' : isWon([res, req]) ? 'lose' : 'draw';
const getRoundTotalScore = ([req, res]: GameRound) =>
  outcomeScore[getRoundOutcome([req, res])] + shapeScore[res];

export const $encryptedRounds = createStore<Encrypted[]>([], { sid: '$encryptedRounds' });
export const $rounds = $encryptedRounds.map(encrypted => encrypted.map(decrypt));
export const $roundsScore = $rounds.map(rounds => rounds.map(getRoundTotalScore));
export const $totalScore = $roundsScore.map(sum);
