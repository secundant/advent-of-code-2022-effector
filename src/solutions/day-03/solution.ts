import { createStore } from 'effector';
import { sum } from '@/shared/math';
import { divideString, getCharCode, isUpperCase, uniq } from '@/shared/std';

export const compartmentsCount = 2;

export const getPriority = (char: string) =>
  isUpperCase(char)
    ? 27 + getCharCode(char) - getCharCode('A')
    : 1 + getCharCode(char) - getCharCode('a');

export const $encryptedRucksacks = createStore<string[]>([], { sid: '$encryptedRucksacks' });

export const $rucksacks = $encryptedRucksacks.map(encrypted =>
  encrypted.map(entry =>
    divideString(entry, entry.length / compartmentsCount).map(part => Array.from(part))
  )
);

export const $rucksacksErrors = $rucksacks.map(rucksacks =>
  rucksacks
    .map(([firstCompartment, ...otherCompartments]) =>
      firstCompartment.filter(itemCode =>
        otherCompartments.some(compartment => compartment.includes(itemCode))
      )
    )
    .map(uniq)
);

export const $failedItemsTotalPriority = $rucksacksErrors
  .map(rucksacks => rucksacks.map(rucksack => rucksack.map(getPriority)).map(sum))
  .map(sum);
