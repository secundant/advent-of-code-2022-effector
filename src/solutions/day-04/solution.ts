import { createStore } from 'effector';
import { toInt } from '@/shared/std';

type SerializedPair = [string, string];
type Range = [number, number];

const parseRange = (input: string) => input.split('-').map(toInt) as [number, number];
const parsePair = (pair: SerializedPair) => pair.map(parseRange) as [Range, Range];

const isRangeContains = (range: Range, subRange: Range) =>
  range[0] <= subRange[0] && range[1] >= subRange[1];

const isFullyOverlappedPair = ([left, right]: [Range, Range]) =>
  isRangeContains(left, right) || isRangeContains(right, left);

export const $serializedPairs = createStore<SerializedPair[]>([], { sid: '$serializedPairs' });

export const $pairs = $serializedPairs.map(pairs => pairs.map(parsePair));

export const $fullyOverlappedPairs = $pairs.map(pairs => pairs.filter(isFullyOverlappedPair));
