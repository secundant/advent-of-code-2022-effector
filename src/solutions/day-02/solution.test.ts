import { fork } from 'effector';
import { describe, expect, test } from 'vitest';
import { readRelativeFile } from '@/shared/fs';
import { $encryptedRounds, $totalScore } from '@/solutions/day-02/solution';

describe('Day 2', () => {
  const createScope = (data: string) =>
    fork({
      values: [
        [
          $encryptedRounds,
          data
            .trim()
            .split('\n')
            .map(line => line.split(' '))
        ]
      ]
    });
  const createScopeWithData = async () =>
    createScope(await readRelativeFile(import.meta, './data.txt'));

  test('should pass example', () => {
    expect(
      createScope(`A Y
B X
C Z`).getState($totalScore)
    ).toBe(15);
  });

  test('01 - total score be if everything goes exactly according to your strategy guide', async () => {
    const scope = await createScopeWithData();

    expect(scope.getState($totalScore)).toMatchSnapshot();
  });
});
