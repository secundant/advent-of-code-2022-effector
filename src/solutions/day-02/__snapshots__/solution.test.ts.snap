// Vitest Snapshot v1

exports[`Day 2 > 01 - total score be if everything goes exactly according to your strategy guide 1`] = `13675`;
